support/elasticsearch_search.erl

```diff
@@ -129,21 +130,27 @@ id_to_integer(Item) ->
 %% @doc Build Elasticsearch query from Zotonic query
 -spec build_query(binary(), {pos_integer(), pos_integer()}, z:context()) -> map().
 build_query(Query, {From, Size}, Context) ->
+    FunctionScore = #{
+        <<"query">> => #{
+            <<"bool">> => #{
+                <<"must">> => lists:filtermap(fun(Q) -> map_query(Q, Context) end, Query),
+                <<"filter">> => build_filter(Query, Context)
+            }
+        },
+        <<"functions">> => lists:filtermap(fun(Q) -> map_score_function(Query, Q, Context) end, Query)
+    },
+    % Because the bool query returns scores of 0, we set the boost_mode to additive instead of multiplicative,
+    % as suggested in https://github.com/elastic/elasticsearch/issues/18273#issuecomment-218482493
+    FunctionScore1 =
+        case proplists:get_value(sort, Query, undefined) of
+            <<"random">> -> FunctionScore#{<<"boost_mode">> => <<"sum">>};
+            _ -> FunctionScore
+        end,
```

Comment:
For clarity, I prefer to keep the structure of the query at the top of this `build_query` function, with function calls nested in that to fill in details. So I would suggest to move the above into a separate function.

Reply:
Agreed

Reply:
Factored the function_score building out.